### ISSUE TITLE
Plugin Fails on Windows. An internal error has occurred. 

### ISSUE THREAD

**Issue by trenzterra** [musselwhizzle/Focus-Points#11]
Hi,

I have D7200 and Lightroom and tried installing this plugin. However, whenever I try to view focus point or show metadata, I get this error:

An internal error has occurred. Couldn't open file "C:\users\username\pictures\dsc_2714-metadata.txt" for reading

This happens with all my NEF files. Is there something I'm doing wrong?

**Comment by musselwhizzle** [comment]
Hi, 
You're not doing anything wrong. I use OSX and dont have a windows machine or LR for windows.  I'm guessing that's a space or a character in the path it doesn't like. If you can figure it out, awesome! If not, give me a little time and within a week of so I have it working. Thanks for the bug report and sorry for the hassle.

**Comment by trenzterra** [comment]
Ah, I see. I can't seem to figure it out though, removing the underscore doesn't help. Thanks a lot for the plugin anyway! It's really useful.

**Comment by musselwhizzle** [comment]
can you give me a sample of what your username is? It doesn't have to be your real name. Like  

"C:\users\Bob's Comp\pictures\dsc_2714-metadata.txt"

**Comment by trenzterra** [comment]
My username is simply Reuben. 

It seems that the plug in is not generating the metadata.txt file at all.

**Comment by jandhollander** [comment]
I have a Canon 80D and I would like to add support for my camera in your plugin :)

It appears I have the same problem. (and yes, I have spaces in the path)

Here is the log I get:
```
01/05/2017 13:55:56 WARN	writing metadata file to: D:\Pictures\2017\2017-01-03 Warhammer\IMG_0623-metadata.txt 
01/05/2017 13:55:56 WARN	exif command: 'D:\Pictures\Lightroom\Plugins\focuspoints.lrdevplugin\bin\exiftool\exiftool' -a -u -g1 'D:\Pictures\2017\2017-01-03 Warhammer\IMG_0623.CR2' > 'D:\Pictures\2017\2017-01-03 Warhammer\IMG_0623-metadata.txt' 
```

**Comment by WimDeMeester** [comment]
@jandhollander I also have a Canon 80D. So if you have something I can test, please let me know!

**Comment by trenzterra** [comment]
Hi, this is the error I get:

`01/05/2017 21:44:07 WARN	starting plugin 
01/05/2017 21:44:07 WARN	writing metadata file to: C:\Users\Reuben\Downloads\test-metadata.txt 
01/05/2017 21:44:07 WARN	exif command: 'D:\LightroomPlugins\Focus-Points-feature-windows_path\focuspoints.lrdevplugin\bin\exiftool\exiftool' -a -u -g1 'C:\Users\Reuben\Downloads\test.nef' > 'C:\Users\Reuben\Downloads\test-metadata.txt' 
`

**Comment by antithesis0** [comment]
I'm having a similar issue, and I think I figured it out:

bin\exiftool\exiftool is a perl script. Perl is not included with windows, (I don't currently have it installed on my windows machine), and it fails quietly.
Let me know if you want help debugging / testing possible solutions.

And thanks for putting working through these issues with the plugin, it's awesome!

**Comment by jandhollander** [comment]
Using 'exiftool.exe' instead of 'exiftool' doesn't fix the problem.

Executing the same command in powershell does write the metadata to the correct file... 
Conclusion: something wrong with executing the exiftool-command on windows?

**Comment by musselwhizzle** [comment]
Exiftool I grabbed from here: http://www.sno.phy.queensu.ca/~phil/exiftool/

Would someone download a windows version and play around figuring out the command or version I need to include into the plugin?

**Comment by Knubbel** [comment]
Using the tags "-a -u -g1 [imagefile] > [textfile]" in exiftollgui does create an screenoutput which ends with an error "File not found: [textfile]" on windows. This might be the same as in the plguin.
When I do want to output normal metadata into an textfile I tend to use the tags "-G1 -a -sort -w _metadata.txt" without any ">" in the line.
So I suggest you use "-a -u -g1 -sort -w _metadata.txt" as tags (with the filenames added where required).

**Comment by d6j6bt** [comment]
I am having the same problem. Installed plug in and every time I try to use it with my D7200 images I get this error.

![focus point error](https://cloud.githubusercontent.com/assets/24942240/21691082/1bcb93d4-d345-11e6-8852-c3f27fc9adee.PNG)
Any help appreciated.

**Comment by musselwhizzle** [comment]
> "-G1 -a -sort -w _metadata.txt" without any ">" in the line

I think that might be it! I'm piping the output to a file which apparently isn't windows friendly. Let me (or someone) run with this idea.

**Comment by Knubbel** [comment]
You just need to use the tag "-w _metadata.txt" to create the file. No piping needed at all. See http://www.sno.phy.queensu.ca/~phil/exiftool/exiftool_pod.html
You do need to append the actual photo file name afterwards: "-W _metadata.txt IMAGEFILENAME"
Maybe the leading minus sign you used for "-metadata.txt" might cause some troubles as exiftool considers a minus to be hinting a a tag.

**Comment by musselwhizzle** [comment]
Ok will work on it within a couple of day (unless you'd like to take a stab at it sooner :) )
https://github.com/musselwhizzle/Focus-Points/tree/feature/windows_path

If you do, note that I shamefully have duplicate methods
https://github.com/musselwhizzle/Focus-Points/issues/14
so watch out for that.

**Comment by jandhollander** [comment]
I'm working on it. Piping works fine in Windows.

I'll keep you posted.

**Comment by jandhollander** [comment]
Fixed and created pull request :)

**Comment by Knubbel** [comment]
Good, using the branch windows-fix the error is gone now and a window opens displaying the image. Only: No AF point is shown (no red rectangle anywhere). Tried it on Canon 80D and Pentax K-1 files.

**Comment by musselwhizzle** [comment]
https://github.com/musselwhizzle/Focus-Points/tree/windows_fix 
has the merged in changes. Sorry for the branch rename. 
@Knubbel - mind trying it on these files which I know works for me on OSX
https://github.com/musselwhizzle/Focus-Points-Sample-Images

**Comment by Knubbel** [comment]
Tried that windows-fix branch on photo DSC_4720 (I had to convert it from NEF to DNG due to my 5.71 LR). Displays the image but no focus point indicator (unless I miss something; I assume it should display right away with no need to press any buttons).

**Comment by antithesis0** [comment]
I also tried the windows-fix branch.

Also got nothing showing except for a preview of the image.

Also, clicking plugin extra -> show metadata, a window loads but all it has is:
"Will place label here" "Will place data here" (no quotes), and nothing else from metadata is showing

**Comment by rderimay** [comment]
Shall I look on it ?

**Comment by musselwhizzle** [comment]
please do if you have the ability. Any issues in "Issues" you'd like to tackle is all you

**Comment by rderimay** [comment]
Ok. I have to rollback, I forgot that I have no LR license on Windows...

Did someone tried to .execute "/bin/exiftool/exiftool(-k).exe" instead after adding the exe ?


On 6 janvier 2017 at 00:07:05, Joshua Musselwhite (notifications@github.com) wrote:

please do if you have the ability. Any issues in "Issues" you'd like to tackle is all you

**Comment by antithesis0** [comment]
Running everything from cmd.exe:
exiftool -G1 -a -sort -w _metadata.txt FILENAME
(no brackets anywhere) seemed to produce the right metadata in .txt

but
exiftool "-G1 -a -sort -w _metadata.txt" FILENAME
fails

Not sure what you mean with the (-k), '"exiftool(-k).exe"' is not recognized as an internal or external command,

**Comment by musselwhizzle** [comment]
> Not sure what you mean with the (-k), '"exiftool(-k).exe"' is not recognized as an internal or external command

For whatever reason, if you download the windows standalone exe it's named. exiftool(-k).exe. The install directions here http://www.sno.phy.queensu.ca/~phil/exiftool/install.html mention to rename it. 

this produces an output for me
```
C:\Users\Joshua\Desktop>"exiftool(-k).exe" -a -sort C:\Users\Joshua\Desktop\DSC-4636.NEF >C:\Users\Joshua\Desktop\metainfo.txt
```
but it tells me to ```-- press RETURN -- ``` to exit the program. Maybe I need to include this windows version in the plugin

**Comment by musselwhizzle** [comment]
It's possible this is the answer, can someone try this: 
https://app.assembla.com/wiki/show/lrdevplugin/Mac_and_Windows_Differences

**Comment by musselwhizzle** [comment]
With the below code, I get the metadata generated and it read in correctly. However, the focus point box never displays and line breaks don't work in "Show metadata" panel. 

```
ExifUtils = {}
exiftool = LrPathUtils.child( _PLUGIN.path, "bin" )
exiftool = LrPathUtils.child(exiftool, "exiftool")
exiftool = LrPathUtils.child(exiftool, "exiftool")

exiftoolWindows = LrPathUtils.child( _PLUGIN.path, "bin" )
exiftoolWindows = LrPathUtils.child(exiftoolWindows, "exiftool.exe")


function ExifUtils.getExifCmd(targetPhoto) 
  
  local path = targetPhoto:getRawMetadata("path")
  local metaDataFile = LrPathUtils.removeExtension(path)
  metaDataFile = metaDataFile .. "-metadata.txt"
  
  local cmd = "'"..exiftool .. "' -a -u -sort '" .. path .. "' > '" .. metaDataFile .. "'";
  if (WIN_ENV) then
    cmd = "\"" .. exiftoolWindows .. " -a -u -sort " .. path .. " > " .. metaDataFile .. "\"";
  end
  log("cmd: " .. cmd)
  log("metaDataFile: " .. metaDataFile) 
  log("-------")
  return cmd, metaDataFile
  
end
```

**Comment by rderimay** [comment]
Yes, that's what I meant yesterday. I think you should include exiftool.exe and call it using exiftool.exe instead of just exiftool like on mac.

For the line breaks not being recognised, maybe this is due to a difference on how windows handles it. This may be better to use the following to iterate through the lines instead of string.match ??
```
local file = io.open("c:\\file.txt")
local tbllines = {}
local i = 0
if file then
    for line in file:lines() do
     i = i + 1
     tbllines[i] = line
    end
    file:close()
end
```

**Comment by jandhollander** [comment]
In the windows_fix branch, exiftool.exe is included.

Generating the metadata.txt files and reading them works fine in this branch, only the metadata doesn't show in the MetadataDialog and the Focus Point doesn't show in the FocusPointDialog.

**Comment by rderimay** [comment]
Yes I see, but this maybe just because of the readIntoTable algorithm not being appropriate for win specifications. Thus falling back to a standard lua core iterator file:lines() "should" be a better idea.

**Comment by jandhollander** [comment]
I think this issue can be closed and 2 separate ones need to be opened for:
- Windows version not showing data in MetaDataDialog
- Windows version not showing Focus point in FocusPointDialog

**Comment by Knubbel** [comment]
I just want to suggest that maybe someone who understands code takes a look at another open source plugin using exiftool: https://github.com/flingo64/PhotoStation-Upload-Lr-Plugin/blob/master/PhotoStation_upload.lrplugin/PSExiftoolAPI.lua
Line #126 reads "parse an exiftool response for a given tag". Maybe the approach chosen there can help here as well?

**Comment by jandhollander** [comment]
I'm working on the MetaDataDialog.

Passing the column1 and column2 texts at creation time, the dialog displays the first line of both texts.

My guess is that we need a carriage return in addition to the line break that is present after each line in the metadata.txt.

I'll keep you posted.

**Comment by musselwhizzle** [comment]
The issue is not with the parsing or writing of the metadata. If you literally pass ``` column1 = "Hello \n World" ``` there is no line break in the text UI. I have no idea how to get it to break. Can anyone figure it out with static/baked text to begin with?

**Comment by musselwhizzle** [comment]
If I had to guess viewFactory:static_text is a singleline view on windows and a multiline view on mac. As far as the focus point not showing, I'm guessing but it seems like on windows the z-index is reversed. It almost worked when I change the order, but it clips the original photo hard. See screenshot.

<img width="1149" alt="screen shot 2017-01-06 at 12 41 57 pm" src="https://cloud.githubusercontent.com/assets/2308674/21732381/0d487682-d40e-11e6-9628-57846623e3d0.png">

**Comment by LowellMontgomery** [comment]
> Ok. I have to rollback, I forgot that I have no LR license on Windows...

You can download a free (30-day) trial for Windows, I suspect.

**Comment by rderimay** [comment]
I'm installing a Windows VM currently in fact.

Could someone summarise what we know about the bug and what have been tried already ?


On 7 janvier 2017 at 19:47:10, Lowell Montgomery (notifications@github.com) wrote:

Ok. I have to rollback, I forgot that I have no LR license on Windows...

You can download a free (30-day) trial for Windows, I suspect.

**Comment by musselwhizzle** [comment]
Let me push up my version to give you a good start. I have to run an errand and can give a summary when I return.

**Comment by LowellMontgomery** [comment]
It is really frustrating that some of the ViewFactory stuff fails only on one platform. Adobe should do a better job of testing the Lightroom SDK for potential cross-platform issues. I've certainly run into some of these, already. (Here, it may be the calls to exiftool, not the LR SDK, which need to be fixed, but…)

**Comment by musselwhizzle** [comment]
https://github.com/musselwhizzle/Focus-Points/tree/feature/josh_windows

I have to run, but currently here are the issues. 

1. viewFactory:static_text seems to be a single line textview on windows. On mac it's multiline and can have line breaks. A simple test of ```"Hello \n World"``` wont even break on windows and demonstrates the issue. This is why the metadata viewer isn't working. 
2. As show in the screenshot, putting the focus point on windows clips the image behind it. I have no idea why. 



<img width="971" alt="screen shot 2017-01-07 at 10 57 52 am" src="https://cloud.githubusercontent.com/assets/2308674/21744330/2c3eac54-d4c8-11e6-868d-d1e60995bf37.png">

**Comment by Knubbel** [comment]
That latest josh_windows version doesnt get to displaying anything on my PC. It fails with the original error about no ....-metadata.txt.

**Comment by musselwhizzle** [comment]
It's possible you have spaces which I mentioned I didn't account for yet

**Comment by LowellMontgomery** [comment]
> viewFactory:static_text seems to be a single line textview on windows.

Bear in mind that there are other options available for viewFactory:static_text, which might be helpful for managing lines, etc, e.g. (where someText is a text string):

```lua
    viewFactory:static_text {
        title = someText,
        width_in_chars = 50,
        wrap = true,
        height_in_lines = 3
    },
```

Using the width_in_chars, wrap, and/or height_in_lines options *might* help.

**Comment by rderimay** [comment]
ok found the reason

**Comment by rderimay** [comment]
ah ! @LowellMontgomery just found it at the same time.

    allow_newlines = true,
    height_in_lines = 20,

is working for me

**Comment by rderimay** [comment]
Btw, height_in_lines is written as readonly in the sdk docs and wraps described just for edit fields... Lovely SDK Documentation

**Comment by LowellMontgomery** [comment]
Yeah, gotta love the SDK docs 😜 
(Not!)

**Comment by rderimay** [comment]
crazy really.

**Comment by rderimay** [comment]
I PR #41 a fix for the metadata dialog. 
Its not perfect but it is working AND aligned (that was a tricky one...)
Still there are some bad chars replaced by a question mark. Pretty sure someone will fix it :-)
Time for sleep

**Comment by rderimay** [comment]
And it still works on Mac, which is crazy :-)

**Comment by rderimay** [comment]
For the af box margins being draw in plain grey, I have done a lot of tries and I have no idea where it could come from.
The fact is if you set background_color = LrColor("blue"), it is blue and not grey any more, which is fine.
But when you set background_color = LrColor() which is transparent, then it becomes grey again and it is not transparent. I suspect it to be a bug when a view is drawn over a catalog_photo...
If still open, I will do more tests tomorrow.

**Comment by musselwhizzle** [comment]
after not touch his code for 3 years, the other dev of the plugin just released an update to his plugin. I guess the multitude of photography blogs writing about us convinced him....Anyway, you can see if his code works for Windows. lightroomfocuspointsplugin.com. If not, then maybe that's an indication it's a stopping point.

**Comment by LowellMontgomery** [comment]
> Still there are some bad chars replaced by a question mark.

Be sure you are saving your code with UTF-8 encoding. That sounds to me like the file may have been saved with (possibly) Windows encoding. If such a change took place, you should be able to see it in the Git diff (at for the character changes).

**Comment by musselwhizzle** [comment]
> Still there are some bad chars replaced by a question mark.
My code sanitizes the metadata as LR text view only supports UTF-8. The code strips all none UTF-8 chars and replaces it with a ? mark. Otherwise LR wont show anything at all.

**Comment by LowellMontgomery** [comment]
Ah... okay. So any non-UTF-8 characters are in the metadata, I get it. :wink:

**Comment by Knubbel** [comment]
@musselwhizzle Since I used the same files yesterday as before I am still suprised that the old error comes up about metadata. Even if it was about blanks/spaces in filenames, these were no issue for one of the paths here  two days ago.
As for the other plug-in: it works on windows, but is making use of the free ImageMagick software to draw images (main reason why the plugin is 14 MB). As he mentions a reference to preview code from Rob Cole this whole display part might be based on Rob Coles excellent work, who had created a framework to use mogrify.exe. Rob Cole delivered his fantastic stuff uncompiled, so you might want to check if any of you have access to the once free "Exportant" plug-in.

**Comment by rderimay** [comment]
@Knubbel I am working an a new Metadata dialog which works now on Windows on my computer. I will push the branch in a few hours so you can test it.
For what it goes about the other plugin, I dont think this is a good idea anyway to base it on imagemagick because you can not have any interaction with the drawn object in the future. In our case, the af-points ar lr objects which can be clicked should we want to in the future.
And the other plugin does not support any sort of crop rotation afaik which was the reason why I uninstalled it a few years ago...

**Comment by rderimay** [comment]
@musselwhizzle @Knubbel 
I have PR a new version that solve a majority a problems on windows : 
It is PR#1 rderimay/new-metadata-dialog.
No idea why it does not show up here, probably because I created the branch myself. Hope you can merge it, otherwise let me know...

Show Metadata (should) work(s) on both platform now, including with spaces in paths. 
Remaining is the fact that is seems that the z-index order is inverted on windows.
I can not really believe this as it is really a breaking difference that would hit everyone.
Much probably we are doing something wrong with the overlapping, but I cannot find what right now...

**Comment by musselwhizzle** [comment]
@rderimay - once you push to your own fork, you need to open a PR. Once a PR is open any pushes you push automatically show up in that PR. If a PR is not already open, you must open one to get your changes into the master repo. 
---
It's looking like to actually get windows support we'll have to do the imagemagick/GD trick. We could potential make a WindowPointsRenderer which returns back a flattened down image so for future expanding Mac could still have click features. It's more work than I have the ability in my schedule. So if @rderimay isn't happen with his results, I'll close this ticket and open one for windows support doc'ing all of our findings. If someone comes along sometimes and wants to work on it, awesome.

**Comment by Knubbel** [comment]
Can you guys then make available one package which I could download and test?

And can you tell me, which script exactly and what part of it is not working about the AF point display with clipping the background? I am happy to try to find someone maybe on other forums who has any further ideas once the problem area can be pinpointed somehow, so I could cite the piece of source.

**Comment by musselwhizzle** [comment]
You can use 
https://github.com/musselwhizzle/Focus-Points/tree/feature/josh_windows
or
https://github.com/musselwhizzle/Focus-Points/tree/feature/rderimay_windows

Good luck :)

**Comment by Knubbel** [comment]
Sorry, but something is going backwards here. I do get the 3 day old error on windows again, which was solved before:
"An internal error has occurred. Couldn't open file "...-metadata.txt" for reading".
This happens with both packages mentioned above.
And I put both the plugin in a path without any blanks in it and put the image itself in C: root, so that can not be the cause. 
Obviously I can not proceed to test any of the nice newer stuff you have done.

**Comment by musselwhizzle** [comment]
You clearly have a space in the path to the image.

**Comment by Knubbel** [comment]
What do you mean? The file path is "C:\DSC_4720.DNG" There is no space.

On another thought about the weird AF point displaying you described:
Have you tested PNGs with different bit depths (8 vs 24)? Or GIFs with transparency?

**Comment by musselwhizzle** [comment]
No space anywhere: not in the path to lightroom, not in the path to the image, not in the path to the plugin? 

This issue isn't the stopping point for windows. So right now, the path thing isn't my biggest concern. If we get the clipping issue working on windows, I'll get the paths working. 

I haven't tested different bits. The images are as shot from the images in raws.

**Comment by Knubbel** [comment]
No spaces in the path to the image or plugin, just in the path to lightroom (C:\Program Files\Adobe\Adobe Photoshop Lightroom 5.7.1) and obviously I can not change "Program files" as it is a core system folder. The thing is that this was working fine 3 days ago as the error was gone even with all sorts of spaces in the path to the image. And I can not test anything else if this error appears beforehand.

**Comment by musselwhizzle** [comment]
```
photo:requestJpegThumbnail( width, height, callback )
```
it's possible instead of display the LrPhoto use the request jpeg and it might not clip. I will try this once I have time.

**Comment by musselwhizzle** [comment]
never mind, it seems like with everything LR on windows, this method is very buggy: 
https://forums.adobe.com/thread/1462400

**Comment by musselwhizzle** [comment]
master branch now has metadata rendering correctly and Metadata viewer should be working fine for windows users as well as mac

**Comment by LowellMontgomery** [comment]
Nice to see so much collaboration on this issue. :smile: For now, I won't keep following it since I have realized, even if I did install Windows and Lightroom to help debug any remaining issues, I'm not able to run the plugin (unsupported camera and no AF lenses at the moment—I'm using a Sony A7R and have bought a couple of the Sony AF lenses, including one with Zeiss elements, and only found the build quality a bit lacking compared to my old/adapted Contax/Zeiss MF lenses, so returned them till I have greater need for AF) :wink:  All the best to you guys. Really nice to see all the Lua/Lightroom SDK collaboration taking place on this project. I'm sure I'll be referring back to this for answers to future frustrations.
